### ISSUE TITLE
Make starting children thread-safe in QProcess

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtbase-os2#63]
There are some things that should be done in a thread-safe way when starting child processes. These include:
1. Changing the working directory.
2. Standard I/O redirection.

If two threads happen to start `QProcess` in parallel, they will easily screw up each other because the current directory and standard I/O streams are process-wide resources shared by all threads.

In LIBCx we provide `spawn2` for exactly this purpose. However, it might not fit well in Qt because Qt does a lot of other things. We might end up with just using a mutex around sensitive parts to provide thread safety. I need to test some things.

**Comment by dmik** [comment]
The above commit actually makes starting children in QProcess completely thread-safe with just adding one single flag to the `spawn2` invocation: P_2_THREADSAFE. This was expected though as this flag was introduced right to provide thread safety (we use it in Python which is used in Mozilla for parallel building and running tests).

The downside of this flag is that it doubles the number of children because there is a wrapper process per each started child (otherwise it's impossible to achieve thread safety on OS/2 because the current directory and file handles are process global resources). And it doesn't cover the `DosStartSession` (detached) case yet as this (P_SESSION support) first needs to be added  in `spawn2` (within the abvove LIBCx ticket).

Doubling child processes is not very good of course (there is a rather tight limit on the number of processes, several hundred or so). But as said above, there is simply no other way to provide complete thread safety. On the other hand, real applications normally do not start many processes and they normally do it from the main thread so that no thread safety is actually necessary. It looks like an overkill to use intermediate wrapper processes in such cases. In order to solve this controversy we may introduce an OS/2-only method to `QProcess`, something like `QProcess::setThreadSafe`. I think this will be pretty legal given that there is a Win32-only enhancement called `QProcess::setCreateProcessArgumentsModifier`.

The only open question is should we be thread-safe by default or not.

**Comment by dmik** [comment]
WIth the closure of #65, all 122 `tst_QProcess` tests run fine now, including the multithreaded ones.

Regarding to the last sentence of the previous comment. `QProcess::setThreadSafe` is there (and used by the tests) but thread safety is disabled by default because it uses a lot of system resources (a wrapper process per each thread-safe start that exists as long as the target process). See `QProcess::setThreadSafe`docs for more info.

Now, given that we're adding P_PM in `spawn2`, I want to enable GUI app tests in `tst_QProcess` (Windows has them).

**Comment by dmik** [comment]
Adding proper GUI app support turned out to be not that fast. I created #66 for that task and will postpone it for now (there is a workaround, see the ticket). This way, there is nothing to do within this task any more. Closing.
